src/spatial_data_format.rs

```diff
@@ -0,0 +1,515 @@
+        let raw_streams = self.get_raw_streams();
+        let interpolated_streams = raw_streams
+            .into_iter()
+            .map(|v| {
+                v.windows(2)
+                    .flat_map(|w| {
+                        let step = (w[1] - w[0]) / samples_per_pt as f32;
+                        (0..samples_per_pt)
+                            .map(|i| w[0] + i as f32 * step)
+                            .collect::<Vec<_>>()
+                    })
+                    .collect()
+            })
+            .collect();
+        Self::attach_tags(&self.header.tags, interpolated_streams)
+    }
+
+    /// Extracts the streams from the file, and quantizes data points to
+    /// produce data points at the requested sample_rate.
+    fn streams_quantized(&self, sample_rate: u64) -> Vec<(GrapeTag, Vec<f32>)> {
+        debug_assert!(sample_rate < self.header.sample_rate);
+        let pts_per_sample = self.header.sample_rate as usize / sample_rate as usize;
+        let raw_streams = self.get_raw_streams();
+        let quantized_streams = raw_streams
+            .into_iter()
+            .map(|v| {
+                v.chunks(pts_per_sample)
+                    .map(|c| c.iter().sum::<f32>() / c.len() as f32)
+                    .collect()
+            })
+            .collect();
+        Self::attach_tags(&self.header.tags, quantized_streams)
+    }
+}
+
+/// This builder contains the data required
+#[derive(Debug, Clone)]
+pub struct GrapeFileBuilder {
+    sample_rate: u64,
+    streams: Vec<(GrapeTag, Vec<f32>)>,
+}
+
+impl Default for GrapeFileBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl GrapeFileBuilder {
+    /// Instaintiate a builder with no streams and a default sample rate of
+    /// 1000 samples per second.
+    fn new() -> Self {
+        GrapeFileBuilder {
+            sample_rate: 1000,
+            streams: Vec::new(),
+        }
+    }
+
+    /// Sets the sample rate of the builder to the argument.
+    pub fn set_samplerate(self, sample_rate: u64) -> Self {
+        GrapeFileBuilder {
+            sample_rate,
+            ..self
+        }
+    }
+
+    /// Adds a tagged stream to the builder.
+    pub fn add_stream(mut self, stream: &[f32], tag: GrapeTag) -> Self {
+        let stream: Vec<f32> = stream.to_vec();
+        self.streams.push((tag, stream));
+        self
+    }
+
+    /// Removes all streams from the builder
+    pub fn clear_streams(mut self) -> Self {
+        self.streams.clear();
+        self
+    }
+
+    /// Builds a [GrapeFile] from the builder, truncating all streams to the
+    /// length of the shortest stream.
+    pub fn build_truncate(self) -> GrapeFile {
+        let tags: Vec<GrapeTag> = self
+            .streams
+            .iter()
+            .map(|(tag, _vec)| tag)
+            .cloned()
+            .collect();
+        let sample_vecs: Vec<Vec<f32>> = self.streams.into_iter().map(|(_tag, vec)| vec).collect();
```

Comment:
After calling `into_iter` we're not suppose to try to access `self.streams` anymore, so that means that after calling this function we shouldn't use the builder anymore right? If that's the case (I may be totally wrong tho) we should mention that in the contract

Reply:
The `build*` family of functions all consume the builder as is the idiom in Rust (see [here](https://rust-unofficial.github.io/patterns/patterns/creational/builder.html)), so a client couldn't reuse the builder even if they wanted to. The entire thing is deallocated because no one owns it anymore.